File: ITB2/src/MinMaxFilter.java
import itb2.filter.AbstractFilter;
import itb2.image.GrayscaleImage;
import itb2.image.Image;
import itb2.image.ImageFactory;

public class MinMaxFilter extends AbstractFilter{
// B = Kreuz aus nur 1, D_B Vierernachbarachscft um Bezugspixel
	private static String MODUS ="modus";
	
	public MinMaxFilter() {
		this.properties.addOptionProperty("modus", "Erosion", "Erosion", "Dilatation");
	}
	
	private double[] get4neighborhood(Image input, int row, int col) {
		return new double[] {input.getValue(col, row-1, GrayscaleImage.GRAYSCALE),
							 input.getValue(col-1, row, GrayscaleImage.GRAYSCALE),
							 input.getValue(col, row, GrayscaleImage.GRAYSCALE),
							 input.getValue(col+1, row, GrayscaleImage.GRAYSCALE),
							 input.getValue(col, row+1, GrayscaleImage.GRAYSCALE)};
	}
	
	private static double getMin(double[] array) {
		double min = array[0];
		for(int i = 0; i<array.length; i++) {
			min=Math.min(min, array[i]);
		}
		return min;
	}
	
	private static double getMax(double[] array) {
		double max = array[0];
		for(int i = 0; i<array.length; i++) {
			max=Math.max(max, array[i]);
		}
		return max;
	}
	
	public Image filter(Image input) {
		String modus = properties.getOptionProperty("modus");
		GrayscaleImage output = ImageFactory.bytePrecision().gray(input.getSize());
		
		int[] DBrow = {-1,0,0,0,1};
		int[] DBcol = {0,-1,0,1,0};
		
		// iteriere über alle Pixel, keine Randbehandlung
		for (int row=1; row<input.getHeight()-1; row++) {
			for (int col=1; col<input.getWidth()-1; col++) {
				double [] subimage = get4neighborhood(input, row, col);
				// +/-B(s,t) ist egal, da konstant 1
				if(modus=="Erosion") {
				output.setValue(col, row, GrayscaleImage.GRAYSCALE, getMin(subimage));
				}
				else {
					output.setValue(col, row, GrayscaleImage.GRAYSCALE, getMax(subimage));
				}
			}	
		}
		return output;
	}
	
}

File: ITB2/src/MinMaxFilter.kt
import itb2.filter.AbstractFilter
import itb2.image.GrayscaleImage
import itb2.image.Image
import itb2.image.ImageFactory
import kotlin.math.floor

class MinMaxFilter : AbstractFilter() {

    /**
     * Replaces values with min / max value of cross shaped environment
     * Converts image to gray scale
     */
    override fun filter(input: Image): Image {
        val grayImage = Grayfilter().filter(input)
        val output: Image = ImageFactory.bytePrecision().gray(input.size)
        val width = grayImage.width
        val height = grayImage.height
        val type = properties.getOptionProperty<String>(TYPE)
        for (col in 0 until width) {
            for (row in 0 until height) {
                val range = 1
                val environment = mutableListOf<Double>()
                for (offsetCol in -range until range + 1) {
                    for (offsetRow in if(offsetCol == 0) 0 until 0 else -range until range + 1) {

                        // only if indices are not beyond the edge getValue is valid
                        if (!(col - offsetCol < 0 || col - offsetCol >= width || row - offsetRow < 0 || row - offsetRow >= height)) {
                            val dif = if(type == "Max") 1 else -1
                            environment.add(grayImage.getValue(col + offsetCol, row + offsetRow, GrayscaleImage.GRAYSCALE) + dif)
                        }
                    }
                }
                output.setValue(col, row, if(type == "Max") environment.maxOrNull() ?: 0.0 else environment.minOrNull() ?: 0.0)
            }
        }
        return output
    }

    companion object {
        private const val TYPE = "type"
    }

    init {
        properties.addOptionProperty(TYPE, "Max", "Max", "Min")
    }

}